frontend/packages/sdk/src/contracts/PinkContract.ts

```diff
@@ -0,0 +1,287 @@
+            result: { ok: { inkMessageReturn: string } };
+          }
+        ).result.ok.inkMessageReturn
+      );
+      return {
+        debugMessage: debugMessage,
+        gasConsumed: gasConsumed,
+        gasRequired: gasRequired && !convertWeight(gasRequired).v1Weight.isZero() ? gasRequired : gasConsumed,
+        output: result.isOk && message.returnType
+          ? this.abi.registry.createTypeUnsafe(message.returnType.lookupName || message.returnType.type, [result.asOk.data.toU8a(true)], { isPedantic: true })
+          : null,
+        result,
+        storageDeposit
+      }
+    }
+
+    return {
+      send: this._decorateMethod((origin: string | AccountId | Uint8Array) => from(asyncMethod(origin)))
+    };
+  };
+
+  #inkCommand = (messageOrId: AbiMessage | string | number, { gasLimit = BN_ZERO, storageDepositLimit = null, value = BN_ZERO }: ContractOptions, params: unknown[]): SubmittableExtrinsic<'promise'> => {
+    const api = this.api as ApiPromise
+
+    // Generate a keypair for encryption
+    // NOTE: each instance only has a pre-generated pair now, it maybe better to generate a new keypair every time encrypting
+    const seed = hexToU8a(hexAddPrefix(randomHex(32)));
+    const pair = sr25519KeypairFromSeed(seed);
+    const [sk, pk] = [pair.slice(0, 64), pair.slice(64)];
+
+    const commandAgreementKey = sr25519Agree(hexToU8a(this.contractKey), sk);
+
+    const asyncMethod = (dest: AccountId, value: BN, gas: { refTime: BN }, storageDepositLimit: BN | undefined, encParams: Uint8Array) => {
+      // @ts-ignore
+      const payload = api.createType("InkCommand", {
+        InkMessage: {
+          nonce: hexAddPrefix(randomHex(32)),
+          // FIXME: unexpected u8a prefix
+          message: api.createType("Vec<u8>", encParams).toHex(),
+          transfer: value,
+          gasLimit: gas.refTime,
+          storageDepositLimit,
+        },
+      });
+      let deposit = new BN(0);
+      const gasFee = new BN(gas.refTime).mul(this.phatRegistry.gasPrice);
+      deposit = new BN(value).add(gasFee).add(new BN(storageDepositLimit || 0));
+      return pinkCommand(api, pk, commandAgreementKey, { contractId: dest.toHex(), payload: payload.toHex(), deposit })
+    }
+
+    return asyncMethod(
+      this.address,
+      // @ts-ignore
+      value,
+      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+      // @ts-ignore jiggle v1 weights, metadata points to latest
+      this._isWeightV1
+        ? convertWeight(gasLimit).v1Weight
+        : convertWeight(gasLimit).v2Weight,
+      storageDepositLimit,
+      this.abi.findMessage(messageOrId).toU8a(params)
+    ).withResultTransform((result: ISubmittableResult) =>
+      // ContractEmitted is the current generation, ContractExecution is the previous generation
+      new ContractSubmittableResult(result, applyOnEvent(result, ['ContractEmitted', 'ContractExecution'], (records: EventRecord[]) =>
```

Comment:
Due to the difference between our execution model and the original ink, we will get no event. The original ink runs the tx on the blockchain. Once the tx is finished, it can filter out the events emitted by the tx call, and return to the user. However on Phala, contracts execute on offchain worker. It can still emit events, but we can only get it from pruntime.

It's good to leave a TODO and move forward for now. However, if we want to support the real event emission, we should:

1. Make sure the emitted event can be retrieved from the worker (pruntime api, or log server)
2. Instead of looking from the `records`, we should wait for the tx execution on worker, and get the event records from it

Reply:
CC @kvinwang

Reply:
https://github.com/Phala-Network/index-contract/pull/42#discussion_r1102239238

Reply:
The `ContractSubmittableResult` will return the block hash, and we can find two on-chain events: `balances.Transfer` and `phalaFatContracts.Transfered`, so I don't think we will get no event here.

To clarify this behavior in the documentation, we should inform users that the events they see only indicate that their transaction has been submitted to the chain. They will still need to fetch the logs from PRuntime to catch any further events that may occur.

My question here: are we logging in PRuntime with the tx code hash?

Reply:
@kvinwang To answer.

Reply:
> are we logging in PRuntime with the tx code hash?

PRuntime doesn't log the events with the tx hash. The tx hash isn't visible there.
There is a `nonce` passed into pruntime, but it is just attached to the MessageOutput and still isn't attached to events in the log.